### PR TITLE
Node enable placeholders

### PR DESCRIPTION
**Pull request by mhoeger**
Resolves https://github.com/Azure/azure-functions-host/issues/4713

### CODE REVIEW

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 1]
```diff
@@ -176,5 +187,36 @@ internal void AddOrUpdateWorkerChannels(RpcWebHostChannelReadyEvent rpcChannelRe
...
+            }
+
+            // Don't use placeholders if running dotnet
+            if (selectedRuntime.Equals(LanguageWorkerConstants.DotNetLanguageWorkerName, StringComparison.InvariantCultureIgnoreCase))
+            {
+                return false;
+            }
+
+            // Check environment config to see if placeholder worker can be used by the function app
+            if (_placeholderEnvironmentConfig.TryGetValue(selectedRuntime, out Dictionary<string, string> environmentConfig))
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: - Move this to separate method
- Add unit tests
- Use this method in GetChannels i.e if condition does not match, return null from GetChannels to ensure, FunctionDispatcher does not get the channel that is about to shutdown

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 2]
```diff
@@ -176,5 +187,36 @@ internal void AddOrUpdateWorkerChannels(RpcWebHostChannelReadyEvent rpcChannelRe
...
+            {
+                foreach (string settingKey in environmentConfig.Keys)
+                {
+                    var currentValue = _environment.GetEnvironmentVariable(settingKey);
+                    environmentConfig.TryGetValue(settingKey, out string cachedValue);
+                    if (!string.Equals(currentValue, cachedValue))
+                    {
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: Use StringComparison.InvariantCultureIgnoreCase

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 3]
```diff
@@ -176,5 +187,36 @@ internal void AddOrUpdateWorkerChannels(RpcWebHostChannelReadyEvent rpcChannelRe
                         return existingLanguageWorkerChannels;
                     });
         }
+
+        public bool MatchingPlaceholderChannelExists(string selectedRuntime)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: Add UnitTests in LanguageWorkerChannelManagerTests

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 4]
```diff
@@ -38,6 +40,15 @@ public WebHostLanguageWorkerChannelManager(IScriptEventManager eventManager, IEn
 
             _shutdownStandbyWorkerChannels = ScheduleShutdownStandbyChannels;
             _shutdownStandbyWorkerChannels = _shutdownStandbyWorkerChannels.Debounce(5000);
+
+            // Set up initial checks for environment variables
+            if (!_environment.IsLinuxHostingEnvironment())
```
[the remark this answers is not in the document]
Comment (mhoeger) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: The environment variable check is specific to windows environment (Linux only runs one version of node and ignores WEBSITE_NODE_DEFAULT_VERSION)

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 5]
```diff
@@ -176,5 +187,36 @@ internal void AddOrUpdateWorkerChannels(RpcWebHostChannelReadyEvent rpcChannelRe
...
+            {
+                return false;
+            }
+
+            // Check environment config to see if placeholder worker can be used by the function app
+            if (_placeholderEnvironmentConfig.TryGetValue(selectedRuntime, out Dictionary<string, string> environmentConfig))
+            {
+                foreach (string settingKey in environmentConfig.Keys)
+                {
+                    var currentValue = _environment.GetEnvironmentVariable(settingKey);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: - What about check for runFromZip? 
- As the matching logic is plarform specific i.e on windows look for specific app setting and on linux ignore, split into two methods

**Comment by mhoeger** [comment]
Closing in favor of https://github.com/Azure/azure-functions-host/pull/4943

**Pull request closed without merging**